Initial API design

fauxnik opened the pull request:
Based loosely on the state reducer pattern. This is how it's designed to be used:

```csharp
var initialState = new MyInitialState(new CommsRadioState(...));
CommsRadioMode.Create(initialState, laserColor);

class MyInitialState : AStateBehaviour
{
  public AStateBehaviour OnAction(InputAction action)
  {
    // Handle action and return new state behaviour class instance
  }

  public AStateBehaviour OnUpdate()
  {
    // Runs on the game loop
    // Handle side-effect-y things here
    // Return a new state behaviour class instance or `this` for no state update
  }
}
```

fauxnik committed:
[WIP] API design

fauxnik committed:
[WIP] Continue API design

Accessor needs to be finished.
CommsRadioController needs to be documented.
	- Is any additional functionality required?
CommsRadioMode is mostly done, but may need additional logging/tweaks.
CommsRadioState is finished & documented.
CommsRadioUtility needs to be documented.
	- Is any additional functionality required?
Main needs work.
	- Needs to hook into and/or patch the base objects.
StateActionUpdateHandler is finished & documented.

fauxnik committed:
[WIP] Add docs site generation via Statiq

fauxnik committed:
[WIP] Update documentation comments

fauxnik committed:
[WIP] Build & deploy docsite w/ github actions

fauxnik committed:
[WIP] Attempt fix build action

fauxnik committed:
Revert "[WIP] Attempt fix build action"

This reverts commit 81296177f5b1c167dba858abba66819f4d26380b.

fauxnik committed:
[WIP] Also checkout submodules

fauxnik committed:
[WIP] Retrieve CommsRadioController & other assets

fauxnik committed:
[WIP] Document CommsRadioMode API

Completes CommsRadioMode.Create setup
Adds a way of retrieving vanilla Comms Radio modes (for use in insertion predicates)

fauxnik committed:
[WIP] Move play sounds to utility & improve docs

Sophon96 commented:
> This is how it's designed to be used:
> 
> ```cs
> var initialState = new MyInitialState(new CommsRadioState(...));
> CommsRadioMode.Create(initialState, laserColor);
> 
> class MyInitialState : StateActionUpdateHandler
> {
>   public StateActionUpdateHandler OnAction(InputAction action)
>   {
>     // Handle action and return new state/handler class instance
>   }
> 
>   public StateActionUpdateHandler OnUpdate()
>   {
>     // Runs on the game loop
>     // Handle side-effect-y things here
>     // Return  a new state/handler class instance or `this` for no state update
>   }
> }
> ```

Looks good to me!

fauxnik committed:
[WIP] Change name of build/deploy workflow

fauxnik committed:
[WIP] Complete API documentation

Still need to write an example of how to use the API.

fauxnik committed:
Rename StateActionUpdateHandler → AStateBehaviour

fauxnik committed:
Add the ability to fork Comms Radio states

fauxnik committed:
Add usage examples to README

fauxnik committed:
Fix github markdown note block syntax

fauxnik committed:
Fix typo

fauxnik committed:
Make optional arguments truly optional

fauxnik committed:
Properly apply starting state

fauxnik committed:
Correct API assembly path in How To Use > Setup

fauxnik committed:
Supress A/B button sound if state didn't change

fauxnik committed:
Don't replace signal origin if it already exists

fauxnik committed:
Expose signal origin via CommsRadioUtility

fauxnik committed:
Fix typo in API documentation

fauxnik committed:
Special case for money removed sound

fauxnik committed:
Allow coroutines to outlive AStateBehaviour instances

Also provide AStateBehaviour information about the previous/next AStateBehaviour being transitioned from/to.
This gives classes that inherit AStateBehaviour enough information to determine when to stop long running coroutines.
